### PR TITLE
Add google_gkeonprem_bare_metal_cluster resource

### PR DESCRIPTION
**Pull request by naitianliu-google**
If this PR is for Terraform, I acknowledge that I have:

- [x] Searched through the [issue tracker](https://github.com/hashicorp/terraform-provider-google/issues) for an open issue that this either resolves or contributes to, commented on it to claim it, and written "fixes {url}" or "part of {url}" in this PR description. If there were no relevant open issues, I opened one and commented that I would like to work on it (not necessary for very small changes).
- [x] Ensured that all new fields I added that can be set by a user appear in at least one [example](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/mmv1/templates/terraform/examples) (for generated resources) or [third_party test](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/mmv1/third_party/terraform/tests) (for handwritten resources or update tests).
- [x] [Generated Terraform providers](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/docs/content/docs/getting-started/generate-providers.md), and ran [`make test` and `make lint`](https://googlecloudplatform.github.io/magic-modules/docs/getting-started/run-provider-tests/#run-unit-tests) in the generated providers to ensure it passes unit and linter tests.
- [x] [Ran](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/docs/content/docs/getting-started/run-provider-tests.md) relevant acceptance tests using my own Google Cloud project and credentials (If the acceptance tests do not yet pass or you are unable to run them, please let your reviewer know).
- [x] Read the [Release Notes Guide](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/.ci/RELEASE_NOTES_GUIDE.md) before writing my release note below.

**Release Note Template for Downstream PRs (will be copied)**

```release-note:new-resource
`google_gkeonprem_bare_metal_cluster`
```

### CODE REVIEW

File: mmv1/third_party/terraform/tests/resource_gkeonprem_bare_metal_cluster_test.go.erb [hunk 1]
```diff
@@ -0,0 +1,403 @@
...
+package google
+<% unless version == 'ga' -%>
+import (
+	"testing"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
+)
+
+func TestAccGkeonpremBareMetalCluster_bareMetalClusterUpdateBasic(t *testing.T) {
+  // VCR fails to handle batched project services
```
Comment (rileykarson) [inline mmv1/third_party/terraform/tests/resource_gkeonprem_bare_metal_cluster_test.go.erb:11]: Note: Same as in https://github.com/GoogleCloudPlatform/magic-modules/pull/7824 we are gonna want to remove these.

Comment (naitianliu-google) [reply mmv1/third_party/terraform/tests/resource_gkeonprem_bare_metal_cluster_test.go.erb]: Removed. Thanks for the correction!

**Review by ScottSuarez — COMMENTED** [review]
Is there external facing documentation for this API. Its tough to validate this resource model when there is no external documentation

**Comment by naitianliu-google** [comment]
> Is there external facing documentation for this API. Its tough to validate this resource model when there is no external documentation

Yes, here is the API docs https://cloud.devsite.corp.google.com/anthos/clusters/docs/on-prem-api/reference/rest (devsite is corp-only)

File: mmv1/products/gkeonprem/BareMetalCluster.yaml [hunk 2]
```diff
@@ -0,0 +1,1003 @@
...
+        name: "multipleNetworkInterfacesConfig"
+        description: |
+          Configuration for multiple network interfaces.
+        properties:
+          - !ruby/object:Api::Type::Boolean
+            name: "enabled"
+            description: |
+              Whether to enable multiple network interfaces for your pods.
+              When set network_config.advanced_networking is automatically
+              set to true.
```
Comment (ScottSuarez) [inline mmv1/products/gkeonprem/BareMetalCluster.yaml:136]: Could this field be tested?

File: mmv1/products/gkeonprem/BareMetalCluster.yaml [hunk 3]
```diff
@@ -0,0 +1,1003 @@
...
+      - !ruby/object:Api::Type::NestedObject
+        name: "srIovConfig"
+        description: |
+          Configuration for SR-IOV.
+        properties:
+          - !ruby/object:Api::Type::Boolean
+            name: "enabled"
+            description: |
+              Whether to install the SR-IOV operator.
```
Comment (ScottSuarez) [inline mmv1/products/gkeonprem/BareMetalCluster.yaml:145]: Could this field be tested/expressed?

File: mmv1/products/gkeonprem/BareMetalCluster.yaml [hunk 4]
```diff
@@ -0,0 +1,1003 @@
...
+        properties:
+          - !ruby/object:Api::Type::String
+            name: "argument"
+            required: true
+            description: |
+              The argument name as it appears on the API Server command line please make sure to remove the leading dashes.
+          - !ruby/object:Api::Type::String
+            name: "value"
+            required: true
+            description: |
```
Comment (ScottSuarez) [inline mmv1/products/gkeonprem/BareMetalCluster.yaml:246]: tests

File: mmv1/products/gkeonprem/BareMetalCluster.yaml [hunk 5]
```diff
@@ -0,0 +1,1003 @@
...
+                    The addresses that are part of this pool. Each address must be either in the CIDR form (1.2.3.0/24) or range form (1.2.3.1-1.2.3.5).
+                - !ruby/object:Api::Type::Boolean
+                  name: "avoidBuggyIps"
+                  description: |
+                    If true, avoid using IPs ending in .0 or .255.
+                    This avoids buggy consumer devices mistakenly dropping IPv4 traffic for those special IP addresses.
+                - !ruby/object:Api::Type::String
+                  name: "manualAssign"
+                  description: |
+                    If true, prevent IP addresses from being automatically assigned.
```
Comment (ScottSuarez) [inline mmv1/products/gkeonprem/BareMetalCluster.yaml:316]: tests

File: mmv1/products/gkeonprem/BareMetalCluster.yaml [hunk 6]
```diff
@@ -0,0 +1,1003 @@
...
+                          description: |
+                            Value associated with the effect.
+                        - !ruby/object:Api::Type::Enum
+                          name: "effect"
+                          description: |
+                            Specifies the nodes operating system (default: LINUX).
+                          values:
+                            - EFFECT_UNSPECIFIED
+                            - PREFER_NO_SCHEDULE
+                            - NO_EXECUTE
```
Comment (ScottSuarez) [inline mmv1/products/gkeonprem/BareMetalCluster.yaml:378]: tests

File: mmv1/products/gkeonprem/BareMetalCluster.yaml [hunk 7]
```diff
@@ -0,0 +1,1003 @@
...
+                  name: "addresses"
+                  required: true
+                  item_type: Api::Type::String
+                  description: |
+                    The addresses that are part of this pool. Each address must be either in the CIDR form (1.2.3.0/24) or range form (1.2.3.1-1.2.3.5).
+                - !ruby/object:Api::Type::Boolean
+                  name: "avoidBuggyIps"
+                  description: |
+                    If true, avoid using IPs ending in .0 or .255.
+                    This avoids buggy consumer devices mistakenly dropping IPv4 traffic for those special IP addresses.
```
Comment (ScottSuarez) [inline mmv1/products/gkeonprem/BareMetalCluster.yaml:475]: tests

File: mmv1/products/gkeonprem/BareMetalCluster.yaml [hunk 8]
```diff
@@ -0,0 +1,1003 @@
...
+                - !ruby/object:Api::Type::String
+                  name: "manualAssign"
+                  description: |
+                    If true, prevent IP addresses from being automatically assigned.
```
Comment (ScottSuarez) [inline mmv1/products/gkeonprem/BareMetalCluster.yaml:479]: tests

File: mmv1/products/gkeonprem/BareMetalCluster.yaml [hunk 9]
```diff
@@ -0,0 +1,1003 @@
...
+                          description: |
+                            Value associated with the effect.
+                        - !ruby/object:Api::Type::Enum
+                          name: "effect"
+                          description: |
+                            Specifies the nodes operating system (default: LINUX).
+                          values:
+                            - EFFECT_UNSPECIFIED
+                            - PREFER_NO_SCHEDULE
+                            - NO_EXECUTE
```
Comment (ScottSuarez) [inline mmv1/products/gkeonprem/BareMetalCluster.yaml:540]: tests

File: mmv1/products/gkeonprem/BareMetalCluster.yaml [hunk 10]
```diff
@@ -0,0 +1,1003 @@
...
+                          Defaults to 10.
+                        default_value: 10
+                      - !ruby/object:Api::Type::Boolean
+                        name: "serializeImagePullsDisabled"
+                        description: |
+                          Prevents the Kubelet from pulling multiple images at a time.
+                          We recommend *not* changing the default value on nodes that run docker
+                          daemon with version  < 1.9 or an Another Union File System (Aufs) storage
+                          backend. Issue https://github.com/kubernetes/kubernetes/issues/10959 has
+                          more details.
```
Comment (ScottSuarez) [inline mmv1/products/gkeonprem/BareMetalCluster.yaml:646]: tests

File: mmv1/products/gkeonprem/BareMetalCluster.yaml [hunk 11]
```diff
@@ -0,0 +1,1003 @@
...
+          Specifies the address of your proxy server.
+          Examples: http://domain
+          WARNING: Do not provide credentials in the format
+          http://(username:password@)domain these will be rejected by the server.
+      - !ruby/object:Api::Type::Array
+        name: "noProxy"
+        item_type: Api::Type::String
+        description: |
+          A list of IPs, hostnames, and domains that should skip the proxy.
+          Examples: ["127.0.0.1", "example.com", ".corp", "localhost"].
```
Comment (ScottSuarez) [inline mmv1/products/gkeonprem/BareMetalCluster.yaml:717]: tests

File: mmv1/products/gkeonprem/BareMetalCluster.yaml [hunk 12]
```diff
@@ -0,0 +1,1003 @@
...
+    properties:
+      - !ruby/object:Api::Type::Array
+        name: "maintenanceAddressCidrBlocks"
+        required: true
+        item_type: Api::Type::String
+        description: |
+          All IPv4 address from these ranges will be placed into maintenance mode.
+          Nodes in maintenance mode will be cordoned and drained. When both of these
+          are true, the "baremetal.cluster.gke.io/maintenance" annotation will be set
+          on the node resource.
```
Comment (ScottSuarez) [inline mmv1/products/gkeonprem/BareMetalCluster.yaml:740]: tests

File: mmv1/products/gkeonprem/BareMetalCluster.yaml [hunk 13]
```diff
@@ -0,0 +1,1003 @@
...
+        default_from_api: true
+      - !ruby/object:Api::Type::Enum
+        name: "containerRuntime"
+        description: |
+          The available runtimes that can be used to run containers in a Bare Metal User Cluster.
+        values:
+          - CONTAINER_RUNTIME_UNSPECIFIED
+          - DOCKER
+          - CONTAINERD
+        default_from_api: true
```
Comment (ScottSuarez) [inline mmv1/products/gkeonprem/BareMetalCluster.yaml:760]: tests

File: mmv1/products/gkeonprem/BareMetalCluster.yaml [hunk 14]
```diff
@@ -0,0 +1,1003 @@
...
+    name: "osEnvironmentConfig"
+    description: |
+      OS environment related configurations.
+    properties:
+      - !ruby/object:Api::Type::Boolean
+        name: "packageRepoExcluded"
+        required: true
+        description: |
+          Whether the package repo should not be included when initializing
+          bare metal machines.
```
Comment (ScottSuarez) [inline mmv1/products/gkeonprem/BareMetalCluster.yaml:783]: tests

File: mmv1/products/gkeonprem/BareMetalCluster.yaml [hunk 15]
```diff
@@ -0,0 +1,1003 @@
...
+            required: true
+            description: Users that will be granted the cluster-admin role on the
+              cluster, providing full access to the cluster.
+            item_type: !ruby/object:Api::Type::NestedObject
+              properties:
+                - !ruby/object:Api::Type::String
+                  name: "username"
+                  required: true
+                  description: |
+                    The name of the user, e.g. `my-gcp-id@gmail.com`.
```
Comment (ScottSuarez) [inline mmv1/products/gkeonprem/BareMetalCluster.yaml:805]: tests

**Comment by ScottSuarez** [comment]
> cloud.devsite.corp.google.com/anthos/clusters/docs/on-prem-api/reference/rest

Any reason for internal only docs? Why not externally available ?

**Review by ScottSuarez — CHANGES_REQUESTED** [review]
Asking for tests on fields that don't seem to have associated examples. If a test can't be supplied thats ok. Just provide a comment with a reason. As long as they are consciously excluded

File: mmv1/products/gkeonprem/BareMetalCluster.yaml (hunk 2 above)
Comment (naitianliu-google) [reply mmv1/products/gkeonprem/BareMetalCluster.yaml:136]: Added. Thanks for checking!

File: mmv1/products/gkeonprem/BareMetalCluster.yaml [hunk 16]
```diff
@@ -0,0 +1,1003 @@
...
+                      description: |
+                        Value associated with the effect.
+                    - !ruby/object:Api::Type::Enum
+                      name: "effect"
+                      description: |
+                        Specifies the nodes operating system (default: LINUX).
+                      values:
+                        - EFFECT_UNSPECIFIED
+                        - PREFER_NO_SCHEDULE
+                        - NO_EXECUTE
```
[the remark this answers is not in the document]
Comment (naitianliu-google) [reply mmv1/products/gkeonprem/BareMetalCluster.yaml:215]: Added

File: mmv1/products/gkeonprem/BareMetalCluster.yaml [hunk 17]
```diff
@@ -0,0 +1,1003 @@
...
+                  name: "addresses"
+                  required: true
+                  item_type: Api::Type::String
+                  description: |
+                    The addresses that are part of this pool. Each address must be either in the CIDR form (1.2.3.0/24) or range form (1.2.3.1-1.2.3.5).
+                - !ruby/object:Api::Type::Boolean
+                  name: "avoidBuggyIps"
+                  description: |
+                    If true, avoid using IPs ending in .0 or .255.
+                    This avoids buggy consumer devices mistakenly dropping IPv4 traffic for those special IP addresses.
```
[the remark this answers is not in the document]
Comment (naitianliu-google) [reply mmv1/products/gkeonprem/BareMetalCluster.yaml:313]: Added

File: mmv1/products/gkeonprem/BareMetalCluster.yaml (hunk 4 above)
Comment (naitianliu-google) [reply mmv1/products/gkeonprem/BareMetalCluster.yaml]: Added

File: mmv1/products/gkeonprem/BareMetalCluster.yaml (hunk 5 above)
Comment (naitianliu-google) [reply mmv1/products/gkeonprem/BareMetalCluster.yaml]: Added

File: mmv1/products/gkeonprem/BareMetalCluster.yaml (hunk 6 above)
Comment (naitianliu-google) [reply mmv1/products/gkeonprem/BareMetalCluster.yaml:379]: Added

File: mmv1/products/gkeonprem/BareMetalCluster.yaml (hunk 8 above)
Comment (naitianliu-google) [reply mmv1/products/gkeonprem/BareMetalCluster.yaml:480]: Added

File: mmv1/products/gkeonprem/BareMetalCluster.yaml (hunk 9 above)
Comment (naitianliu-google) [reply mmv1/products/gkeonprem/BareMetalCluster.yaml:541]: Added

File: mmv1/products/gkeonprem/BareMetalCluster.yaml (hunk 10 above)
Comment (naitianliu-google) [reply mmv1/products/gkeonprem/BareMetalCluster.yaml]: The subfields "cpuManagerPolicy", "cpuCfsQuotaDisabled", "cpuCfsQuotaPeriod", "featureGates", "podPidsLimit" have been removed from the API surface recently. Removed from TF config in this PR and added rest of subfields in tests

File: mmv1/products/gkeonprem/BareMetalCluster.yaml (hunk 14 above)
Comment (naitianliu-google) [reply mmv1/products/gkeonprem/BareMetalCluster.yaml:723]: Added

File: mmv1/products/gkeonprem/BareMetalCluster.yaml (hunk 7 above)
Comment (naitianliu-google) [reply mmv1/products/gkeonprem/BareMetalCluster.yaml:476]: Added

File: mmv1/products/gkeonprem/BareMetalCluster.yaml (hunk 11 above)
Comment (naitianliu-google) [reply mmv1/products/gkeonprem/BareMetalCluster.yaml:657]: Added

File: mmv1/products/gkeonprem/BareMetalCluster.yaml (hunk 12 above)
Comment (naitianliu-google) [reply mmv1/products/gkeonprem/BareMetalCluster.yaml:680]: Added

File: mmv1/products/gkeonprem/BareMetalCluster.yaml (hunk 13 above)
Comment (naitianliu-google) [reply mmv1/products/gkeonprem/BareMetalCluster.yaml:700]: Added

File: mmv1/products/gkeonprem/BareMetalCluster.yaml (hunk 15 above)
Comment (naitianliu-google) [reply mmv1/products/gkeonprem/BareMetalCluster.yaml:745]: Added

File: mmv1/products/gkeonprem/BareMetalCluster.yaml (hunk 3 above)
Comment (naitianliu-google) [reply mmv1/products/gkeonprem/BareMetalCluster.yaml:145]: Added

**Review by naitianliu-google — COMMENTED** [review]
Thanks for taking time reviewing this PR. Added accordingly by increasing test coverage and fixing the Resource config file based on test results. Please let me know if anything else needs to be updated. Thanks! @ScottSuarez

File: mmv1/products/gkeonprem/BareMetalCluster.yaml (hunk 10 above)
Comment (ScottSuarez) [reply mmv1/products/gkeonprem/BareMetalCluster.yaml]: This makes me slightly nervous. How stable is this API? I worry about adding fields which may be unsupported by the API at a future point

**Comment by naitianliu-google** [comment]
> > cloud.devsite.corp.google.com/anthos/clusters/docs/on-prem-api/reference/rest
> 
> Any reason for internal only docs? Why not externally available ?

Here is the public one which was just released. https://cloud.google.com/anthos/clusters/docs/on-prem-api/reference/rest

File: mmv1/products/gkeonprem/BareMetalCluster.yaml [hunk 18]
```diff
@@ -0,0 +1,943 @@
...
+    name: "osEnvironmentConfig"
+    description: |
+      OS environment related configurations.
+    properties:
+      - !ruby/object:Api::Type::Boolean
+        name: "packageRepoExcluded"
+        required: true
+        description: |
+          Whether the package repo should not be included when initializing
+          bare metal machines.
```
Comment (ScottSuarez) [inline mmv1/products/gkeonprem/BareMetalCluster.yaml:723]: This field doesn't seem to be documented externally
https://cloud.google.com/anthos/clusters/docs/on-prem-api/reference/rest/v1/projects.locations.bareMetalClusters

**Review by ScottSuarez — COMMENTED** [review]
Some fields don't appear to be documented externally or are represented differently.. Could you confirm?

**Comment by naitianliu-google** [comment]
> Some fields don't appear to be documented externally or are represented differently.. Could you confirm?

The external document is to up-to-date. The latest version is not released yet. Please reference the internal one for the latest changes. (Apologize for the incorrect reference. ) https://cloud.devsite.corp.google.com/anthos/clusters/docs/on-prem-api/reference/rest/v1/projects.locations.bareMetalClusters#BareMetalNetworkConfig

File: mmv1/products/gkeonprem/BareMetalCluster.yaml [hunk 19]
```diff
@@ -0,0 +1,943 @@
...
+    name: "controlPlane"
+    description: |
+      Specifies the control plane configuration.
+    required: true
+    properties:
+      - !ruby/object:Api::Type::NestedObject
+        name: "controlPlaneNodePoolConfig"
+        required: true
+        description: |
+          Configures the node pool running the control plane. If specified the corresponding NodePool will be created for the cluster's control plane. The NodePool will have the same name and namespace as the cluster.
```
[the remark this answers is not in the document]
Comment (naitianliu-google) [reply mmv1/products/gkeonprem/BareMetalCluster.yaml:156]: The external document is out of date. Please refer to the internal one. https://cloud.devsite.corp.google.com/anthos/clusters/docs/on-prem-api/reference/rest/v1/projects.locations.bareMetalClusters#baremetalcontrolplaneconfig

File: mmv1/products/gkeonprem/BareMetalCluster.yaml (hunk 18 above)
Comment (naitianliu-google) [reply mmv1/products/gkeonprem/BareMetalCluster.yaml:723]: The external document is out of date. Please refer to the internal one. [https://cloud.devsite.corp.google.com/anthos/clusters/docs/on-prem-api/reference/rest/v1/projects.locations.bareMetalClusters](https://cloud.devsite.corp.google.com/anthos/clusters/docs/on-prem-api/reference/rest/v1/projects.locations.bareMetalClusters#baremetalcontrolplaneconfig)

**Comment by naitianliu-google** [comment]
@ScottSuarez @rileykarson Please help give a pass on this PR when you get a chance. Thanks!

**Review by ScottSuarez — APPROVED** [review]

**Pull request merged**